packages/eslint-plugin/src/rules/prefer-regexp-exec.ts

```diff
@@ -37,25 +46,41 @@ export default createRule({
     const sourceCode = context.getSourceCode();
 
     /**
-     * Check if a given node is a string.
-     * @param node The node to check.
+     * Check if a given node type is a string.
+     * @param node The node type to check.
      */
-    function isStringType(node: TSESTree.Node): boolean {
-      const objectType = typeChecker.getTypeAtLocation(
-        parserServices.esTreeNodeToTSNodeMap.get(node),
-      );
-      return getTypeName(typeChecker, objectType) === 'string';
+    function isStringType(type: ts.Type): boolean {
+      return getTypeName(typeChecker, type) === 'string';
     }
 
     /**
-     * Check if a given node is a RegExp.
-     * @param node The node to check.
+     * Check if a given node type is a RegExp.
+     * @param node The node type to check.
      */
-    function isRegExpType(node: TSESTree.Node): boolean {
-      const objectType = typeChecker.getTypeAtLocation(
-        parserServices.esTreeNodeToTSNodeMap.get(node),
-      );
-      return getTypeName(typeChecker, objectType) === 'RegExp';
+    function isRegExpType(type: ts.Type): boolean {
+      return getTypeName(typeChecker, type) === 'RegExp';
+    }
+
+    function collectArgumentTypes(types: ts.Type[]): ArgumentType {
+      let result = ArgumentType.Other;
+
+      for (const type of types) {
+        if (isRegExpType(type)) {
+          if (result === ArgumentType.Other) {
+            result = ArgumentType.RegExp;
+          } else if (result === ArgumentType.String) {
+            result = ArgumentType.Both;
+          }
+        } else if (isStringType(type)) {
+          if (result === ArgumentType.Other) {
+            result = ArgumentType.String;
+          } else if (result === ArgumentType.RegExp) {
+            result = ArgumentType.Both;
+          }
```

Comment:
> Added line #L78 was not covered by tests

Well, both of these are covered in test cases:

```
const matchers = ['package-lock.json', /regexp/];
const matchers = [/regexp/, 'package-lock.json'];
```

...so I think this is a coincidental implementation detail in how TypeScript happens to sort union types.